### PR TITLE
Querying unflushed-memory-chain

### PR DESCRIPTION
**Pull request by Tushar3099**
Signed-off-by: Tushar3099 <tusharneogi30@gmail.com>

Fixes Issue: #378 

**Type Of Change**
----

- [x] Enhancement

**Changes**
----
- Modified `ExecWithoutEncode` function to append in-memory chain to the chain retrieved from the secondary storage.
- Modified `/query` and `/query-matrix` route handlers to access the in-memory chains and pass it to `ExecWithoutEncode`
- Modified frontend fetch routes accordingly.

**Results**
- Querying in-memory chain in `Monitoring` window and `Test` window.

**Preview**
----

https://user-images.githubusercontent.com/51849085/112421453-0d4fc400-8d55-11eb-865f-96bc8d36ddc7.mp4

### CODE REVIEW

**Comment by Tushar3099** [comment]
@Harkishen-Singh I have accessed the in-memory chain of a particular hash using BRMatrix and then appended the chain to the chain obtained from secondary storage and sent this to the frontend. I assume the result is showing what was expected. 
Please tell me if the approach I used is right.
Both the Testing window and Monitoring window are querying flushed data along with in-memory chains.

File: src/lib/api/api.go [hunk 1]
```diff
@@ -174,8 +174,10 @@ func (a *API) Query(w http.ResponseWriter, r *http.Request) {
 		startTimestamp, endTimestamp int64
 		err                          error
 	)
-	timeSeriesPath := r.URL.Query().Get("timeSeriesPath")
-
+	hash := r.URL.Query().Get("hashKey")
+	Type := r.URL.Query().Get("type")
```
Comment (Harkishen-Singh) [inline src/lib/api/api.go:178]:
```suggestion
	var (
	    hash = r.URL.Query().Get("hashKey")
	    queryType = r.URL.Query().Get("type")
	)
```

File: src/lib/api/api.go [hunk 2]
```diff
@@ -174,8 +174,10 @@ func (a *API) Query(w http.ResponseWriter, r *http.Request) {
...
+	var timeSeriesPath string
+	var newBlocks []tsdb.Block
```
Comment (Harkishen-Singh) [inline src/lib/api/api.go:180]: These two are used down. I suggest declaring them close to their usecase.

File: src/lib/api/api.go [hunk 3]
```diff
@@ -196,25 +198,54 @@ func (a *API) Query(w http.ResponseWriter, r *http.Request) {
...
+			return
+		}
+		matrix := (*a.Matrix)[hash]
+
+		switch Type {
+		case "ping":
+			newBlocks = matrix.PingChain.Chain
+		case "jitter":
+			newBlocks = matrix.JitterChain.Chain
+		default:
```
Comment (Harkishen-Singh) [inline src/lib/api/api.go:228]: Will `Type` not have `monitoring`?

File: src/lib/api/api.go [hunk 4]
```diff
@@ -196,25 +198,54 @@ func (a *API) Query(w http.ResponseWriter, r *http.Request) {
...
+			timeSeriesPath = "storage/journal" + tsdb.FileExtension
+		default:
+			a.send(w, []byte("INVALID_TYPE"))
+			return
+		}
+		if ok := tsdb.VerifyChainPathExists(timeSeriesPath); !ok {
+			a.send(w, []byte("INVALID_PATH"))
+			return
+		}
+		newBlocks = []tsdb.Block{}
```
Comment (Harkishen-Singh) [inline src/lib/api/api.go:215]: Why is this required?

File: src/lib/api/api.go [hunk 5]
```diff
@@ -196,25 +198,54 @@ func (a *API) Query(w http.ResponseWriter, r *http.Request) {
...
-	}
 
 	// TODO: we do not capture the block streams in memory while querying yet. They are captured only when flushed.
 	// TODO: consider cmap while querying for fetching latest blocks after shifting tsdb to binary.
 
 	qry := querier.New(timeSeriesPath, "", querier.TypeRange)
 	query := qry.QueryBuilder()
 	query.SetRange(startTimestamp, endTimestamp)
-	a.Data = query.ExecWithoutEncode()
+	a.Data = query.ExecWithoutEncode(newBlocks)
```
Comment (Harkishen-Singh) [inline src/lib/api/api.go:248]: I see newBlocks is always going to be empty? So is there a reason to pass this down the func?

File: tsdb/querier/querier.go [hunk 6]
```diff
@@ -88,9 +88,10 @@ func (q *Query) Exec() []byte {
 
 // ExecWithoutEncode executes the Query without encoding the result to []byte,
 // rather keeps the result as default QueryResponse.
-func (q *Query) ExecWithoutEncode() QueryResponse {
+func (q *Query) ExecWithoutEncode(newBlocks []tsdb.Block) QueryResponse {
```
Comment (Harkishen-Singh) [inline tsdb/querier/querier.go:91]: If you need blocks to be passed for querying, we should have a `FillDataBlocks()` that should do this job. The param here is not defining anything IMO.

**Review by harkishen — CHANGES_REQUESTED** [review]

File: src/lib/api/api.go (hunk 2 above)
Comment (Tushar3099) [reply src/lib/api/api.go:180]: I will do these changes.

File: src/lib/api/api.go (hunk 4 above)
Comment (Tushar3099) [reply src/lib/api/api.go:215]: Actually, `Query` is used for both `system` information and `tests`(ping/jitter). Incase of jitter/ping I am sending the route hash to access the in-memory chains because through `BRMatrix` we can access the chains, but we cannot access the system information like that so I am just sending the flushed data for system metrics.

File: src/lib/api/api.go (hunk 3 above)
Comment (Tushar3099) [reply src/lib/api/api.go:228]: Because `Query` is used for Testing which only has Ping and Jitter for now, so I just made cases for them. I will include monitoring also for future purposes.

File: src/lib/api/api.go (hunk 5 above)
Comment (Tushar3099) [reply src/lib/api/api.go:248]: No not-neccessary as : 
```
		case "ping":
			newBlocks = matrix.PingChain.Chain
		case "jitter":
			newBlocks = matrix.JitterChain.Chain
		default:
```
`newBlocks` will have the unflushed memory-chains which are monitored and may have new blocks.

File: tsdb/querier/querier.go (hunk 6 above)
Comment (Tushar3099) [reply tsdb/querier/querier.go:91]: Actually, earlier in `ExecWithoutEncode`, we reloaded all the blocks from the DBpath and ran `exec` function. Now I am passing down the newBlocks(in-memory chain), which is then appended to the fetched blocks from the local storage, then we are executing `exec`. Please explain how can I implement this otherwise.

File: src/lib/api/api.go (hunk 5 above)
Comment (Tushar3099) [reply src/lib/api/api.go:248]: I am having this inference from main.go : 
```
		chainSet.Register(fmt.Sprintf("%s-ping", uHash), targetMachineCalc[uHash].Ping)
		chainSet.Register(fmt.Sprintf("%s-jitter", uHash), targetMachineCalc[uHash].Jitter)
		chainSet.Register(fmt.Sprintf("%s-fping", uHash), targetMachineCalc[uHash].FPing)
```
as targetMachineCalc[uHash] are registered in the `chainset` and also,
```
	        matrix[hash] = &utils.BRMatrix{
		        Route:        r,
		        PingChain:    targetMachineCalc[uHash].Ping,
		        JitterChain:  targetMachineCalc[uHash].Jitter,
		        FPingChain:   targetMachineCalc[uHash].FPing,
		        MonitorChain: tsdb.NewChain(pathMonitor).Init(),
		        Metrics:      endpointMetrics,
	        }
```
targetMachineCalc[uHash] is assigned to the matrix , so when : 
```
		        case "Ping":
			        workers.Ping.Iterate("stop", false)
		        case "FloodPing":
			        workers.PingF.Iteratef("stop", false)
		        case "Jitter":
			        workers.Jitter.Iterate("stop", false)
		        case "ReqResDelayMonitoring":
			        workers.Monitor.Iterate("stop", false)
		        }
```
It is used to iterate over all the routes the through `targetMachineCalc[uHash]`, in-memory chains will be updated in matrix as in worker we pass `targetMachineCalc[uHash]`,
```
	workers := &struct {
		Ping    *ping.Ping
		Jitter  *jitter.Jitter
		PingF   *ping.FloodPing
		Monitor *monitor.Monitor
	}{
		Ping:    ping.New(conf, ping.TestInterval{OfType: intervals[0].Type, Duration: *intervals[0].Duration}, &targetMachineCalc),
		Jitter:  jitter.New(conf, jitter.TestInterval{OfType: intervals[0].Type, Duration: *intervals[1].Duration}, &targetMachineCalc),
		PingF:   ping.Newf(conf, ping.TestInterval{OfType: intervals[0].Type, Duration: *intervals[0].Duration}, conf.Config.Password, &targetMachineCalc),
		Monitor: monitor.New(conf, monitor.TestInterval{OfType: intervals[2].Type, Duration: *intervals[2].Duration}, &matrix),
	}
```

**Comment by Tushar3099** [comment]
@Harkishen-Singh, Can you please reply to the comments so that I can improve the PR and make the necessary changes? Thank you.
